pkg/store/bucket.go

```diff
@@ -777,33 +784,132 @@ type seriesEntry struct {
 	chks []storepb.AggrChunk
 }
 
-type bucketSeriesSet struct {
-	set []seriesEntry
-	i   int
-	err error
+// blockSeriesClient is a storepb.Store_SeriesClient for a
+// single TSDB block in object storage.
+type blockSeriesClient struct {
+	grpc.ClientStream
+	ctx     context.Context
+	extLset labels.Labels
+
+	ps             []storage.SeriesRef
+	i              int
+	mint           int64
+	maxt           int64
+	indexr         *bucketIndexReader
+	chunkr         *bucketChunkReader
+	limiter        ChunksLimiter
+	loadAggregates []storepb.Aggr
+	skipChunks     bool
+
+	shardMatcher       *storepb.ShardMatcher
+	calculateChunkHash bool
+
+	// Transform all series into the response types and mark their relevant chunks
+	// for preloading.
+	symbolizedLset []symbolizedLabel
+
+	response *storepb.SeriesResponse
 }
 
-func newBucketSeriesSet(set []seriesEntry) *bucketSeriesSet {
-	return &bucketSeriesSet{
-		set: set,
-		i:   -1,
+func emptyBlockSeriesClient() *blockSeriesClient {
+	return &blockSeriesClient{
+		ps: nil,
 	}
 }
 
-func (s *bucketSeriesSet) Next() bool {
-	if s.i >= len(s.set)-1 {
-		return false
+func newBlockSeriesClient(
+	ctx context.Context,
+	extLset labels.Labels,
+	ps []storage.SeriesRef,
+	minTime int64,
+	maxTime int64,
+	indexr *bucketIndexReader,
+	chunkr *bucketChunkReader,
+	limiter ChunksLimiter,
+	skipChunks bool,
+	loadAggregates []storepb.Aggr,
+	shardMatcher *storepb.ShardMatcher,
+	calculateChunkHash bool,
+) *blockSeriesClient {
+	return &blockSeriesClient{
+		ctx:        ctx,
+		extLset:    extLset,
+		ps:         ps,
+		i:          -1,
+		mint:       minTime,
+		maxt:       maxTime,
+		indexr:     indexr,
+		chunkr:     chunkr,
+		limiter:    limiter,
+		skipChunks: skipChunks,
+
+		loadAggregates:     loadAggregates,
+		shardMatcher:       shardMatcher,
+		calculateChunkHash: calculateChunkHash,
 	}
-	s.i++
-	return true
 }
 
-func (s *bucketSeriesSet) At() (labels.Labels, []storepb.AggrChunk) {
-	return s.set[s.i].lset, s.set[s.i].chks
-}
+func (b *blockSeriesClient) Recv() (*storepb.SeriesResponse, error) {
+	b.i++
+	if b.i >= len(b.ps) {
+		return nil, io.EOF
+	}
+
+	var chks []chunks.Meta
+	ok, err := b.indexr.LoadSeriesForTime(b.ps[b.i], &b.symbolizedLset, &chks, b.skipChunks, b.mint, b.maxt)
```

Comment:
Here we probably want to fetch series in batches instead of one at a time. The size of the batch is hard to determine, but maybe 100 would be a good first try?